### PR TITLE
Add Spawner.auth_state_hook

### PR DESCRIPTION
**Pull request by rcthomas**
Here's an example of what I'm thinking of as a solution for #2552.  Another idea I had was a custom handler, but in #1854 it was agreed not to override existing handlers.  Is there a better way to accomplish this, or, is there a reason it would be bad to directly pass this info to a spawner?

### CODE REVIEW

File: jupyterhub/spawner.py [hunk 1]
```diff
@@ -627,6 +627,10 @@ def my_hook(spawner):
         """
     ).tag(config=True)
 
+    def handle_auth_state(self, auth_state):
+        """Pass optional auth_state to Spawner."""
+        pass
```
Comment (minrk) [inline jupyterhub/spawner.py]: Should there perhaps be a configurable auth_state_hook so this can be used via config without subclassing?

**Comment by minrk** [comment]
My only concern with this is the overlap with pre_spawn_hook, which accomplishes much the same thing since it is called at the same time with the same info available, except that `get_auth_state` must be called in the hook. What do you think?

**Comment by rcthomas** [comment]
I can look again but I think `run_pre_spawn_hook` runs too late for what I'm trying to make happen.  There are things I want to happen before the notebook is spawned that depend on `auth_state` and the spawner seems like the right place to put those.  Maybe there's a better place?

**Comment by vilhelmen** [comment]
My dream that I'm slowly forcing^H^H^H^H^H^H^H suggesting to @minrk is to get `auth_state` everywhere, acting like a user profile for all sorts of runtime configuration, but what we've got now works pretty well. The user object is generally reachable when needed, and a surprising number of configuration settings support callables. Calling `get_auth_state()` in whatever spawner functions may really be best.

Worst case, you could always subclass your spawner, override init, and after calling super call get_auth_state and save the dict to self, unless there's some async funkiness involved.

**Comment by rcthomas** [comment]
Two things in my mind led to this PR.  First I was looking for a way to stick some extra (not even secret) stuff onto `user` to carry around. Specifically, "this user can do this" or "this user can charge to this time allocation."  I overlooked `auth_state` for a while because the name and the encryption made me think it wasn't the appropriate place for that, but looking back at old issues convinced me it was OK.  So that solved that problem, but it made me think the `user` concept needs to be extendable. I see some other issues/comments on roles and things that indicate others have thought about this.

The second thing that led to this implementation was that I needed to `await` that `get_auth_state()` somewhere.  (I looked around at a lot of places to inject it, I guess that is "async funkiness" as you say.)  It might be better to do this in the base handler instead of in just the one method I personally happened to need it in (which was the `get()` for rendering the home page)?

What I've done here isn't that good even because some stuff in `auth_state` could change while the user is off getting coffee, but I think the auth can be refreshed.  That's already a shortcoming of this PR.

**Comment by vilhelmen** [comment]
I'm all for flexibility in the system, I've been wanting to rename `auth_state` to `profile` to make it a little less spooky sounding and get it more actively used as a general "stick user data here for everything else to reference" blob.

You should, in theory, be able to immediately await that `get_auth_state()` call since the `option_form()` callable is awaited in the spawner, but I tend to shy away from the async bits because I end up needing Min to bail me out in my PRs. Multithreaded C? Sure, any time. Async python? Uggghhh.

I've been curious about the auth state refresh system (I believe the function is called something like that), and that may be a way to go, but I've never used it. Not having read the code, it should work just fine with named servers in JH 1.0. I'm not sure if your spawner tracks total session runtime, but the auth state refresh function should be called on a timer roughly often enough to use it as a manual timekeeper.

I've been meaning to propose some user onboarding handler, basically a `new_user_add` callback to act like a non-auth `auth_state` initialization system, but I don't think that'd really help in this case.

File: jupyterhub/spawner.py [hunk 2]
```diff
@@ -953,6 +971,14 @@ def run_post_stop_hook(self):
             except Exception:
                 self.log.exception("post_stop_hook failed with exception: %s", self)
 
+    def run_auth_state_hook(self, auth_state):
+        """Run the auth_state_hook if defined"""
+        if self.auth_state_hook is not None:
+            try:
+                return self.auth_state_hook(self, auth_state)
+            except Exception:
+                self.log.exception("auth_stop_hook failed with exception: %s", self)
```
Comment (willingc) [inline jupyterhub/spawner.py]:
```suggestion
                self.log.exception("auth_state_hook failed with exception: %s", self)
```

File: jupyterhub/spawner.py [hunk 3]
```diff
@@ -627,6 +627,24 @@ def my_hook(spawner):
         """
     ).tag(config=True)
 
+    auth_state_hook = Any(
+        help="""
+        An optional hook function that you can implement to pass `auth_state`
+        to the spawner after it has been initialized but before it starts.
+        The `auth_state` dictionary may be set by the `.authenticate()`
+        method of the authenticator.  This hook enables you to pass some
+        or all of that information to your spawner.
```
Comment (willingc) [inline jupyterhub/spawner.py]:
```suggestion
        or all of the `auth_state` information to your spawner.
```

**Comment by rcthomas** [comment]
I have been running this in production for a while here and it seems to be OK, but I would be happy to learn that there's a better place (or more places) where the hook should be called.

**Comment by minrk** [comment]
Thanks for your patience on this one @rcthomas and thanks @willingc for the review!

**Comment by hvpareja** [comment]
Hi there,

sorry @rcthomas for reopening this threat, but I think the documentation is not helping me to leverage this new feature.

I am very interested to use the auht_state_hook, but I fail to understand where exactly needs to be placed. Could you extend the same example you wrote in the documentation, please?

Thanks

**Pull request merged**